### ISSUE TITLE
Python indentation completely derailed by comments

### ISSUE THREAD

**Issue by yeti-dn** [vim/vim#15404]
### Steps to reproduce

1. Open the attached Python script (which I apparently cannot attach because ‘We don't support that file type’ – see inside the ZIP)

gvim --clean silly-indent.py

The script is almost empty. Almost all content is just a single comment containing excerpts of several text data files the script is supposed to process.

2. Press G (to go to the end), press o (to start insert mode below).

3. Press Enter repeatedly (and slowly).

[silly-indent.zip](https://github.com/user-attachments/files/16441754/silly-indent.zip)

### Expected behaviour

New empty lines are added. The cursor stays in the first column. Adding a new empty line does not cause noticeable delay, at least on adequate hardware.

### Version of Vim

9.1.393

### Environment

Operating system: Fedora 38

I am using the distro vim package. The indentation function should be the same as the latest version.

The hardware is more than adequate, AMD Ryzen 5 PRO 5650G with 64 GB RAM.

### Logs and stack traces

```shell
I am not sure how to get a useful log from python#GetIndent().
A screen recording is attached instead (again inside the ZIP).

Observed behaviour:

New lines are added.
However, each press of Enter causes a noticeable delay.
Furthermore, cursor is jumping left and right wildly upon each Enter.
The indentation seems random, but probably it is in fact deterministic, just strange.
```

**Comment by yeti-dn** [comment]
You are right, on a quiescent system I do not get random indentation with the default timeout either. Under load (there were also some computations running), I get simultaneously the delays and random indentation.

**Comment by yeti-dn** [comment]
So, the problem is that the comment contains lots of parentheses and so the skip lambda

`{-> synstack('.', col('.'))  ->indexof({_, id -> synIDattr(id, 'name') =~ '\%(Comment\|Todo\|String\)$'}) >= 0}`

is evaluated a lot in order to skip them all, right?

When I simply remove the skip argument of searchpairpos() it becomes fast again (but of course it can become confused by parentheses in comments and strings). A workaround is then to set

`pyindent_disable_parentheses_indenting = 1`

as having to occasionally indent parentheses manually is much less frustrating. And I suppose true fix requires having a more efficient way of skipping text in searchpairpos() based on syntax than this…

**Comment by chrisbra** [comment]
I think both `searchpairpos()` is already not fast and in addition, getting the syntax stack for each match is also slow. So I think your workaround makes sense

**Comment by yeti-dn** [comment]
According to my – admittedly anecdotal – testing, removing the skip argument of searchpairpos() makes it snappy, even on a system under load. So here the culprit is really the interpreted code. I cannot imagine how a builtin function alone could spend 150 ms on couple of kB of text, that is ~ 10⁵ instructions per character.

**Comment by chrisbra** [comment]
It's strange, I have starred a bit at the code, but I cannot find out what causes searchpairpos() to return an invalid position even so the timeout should make the search function abort. Not sure what is going on (and I cannot seem to reproduce this under a debugger)

**Comment by yeti-dn** [comment]
I do not think you will find the reason in do_searchpair(). I added some manual printf()s and the strange indent occurs when

`r = eval_expr_to_bool(skip, &err);`

gives 0 inside the comment. Then the function concludes it found something [that is not a comment] and proceeds accordingly.

Now, why does it happen? If I change the `skip` lambda in SearchBracket() to a normal function, making it easier to debug what is going on inside, I sometimes see that synstack() returns an empty stack, leading to `r = 0`. But when it happens, the current row and column are somewhere in the middle of a comment line, so the stack should not be empty.

Indeed, manual evaluation of synstack() at the same location does *not* return an empty stack. What is different inside searchpairpos()?

The comment is somewhere relatively far above. So synstack() probably needs to do parsing there. Can it hit some regexp timeout, hence fail to mark the comment and return an empty stack?

**Comment by yeti-dn** [comment]
A cleaner test case. Make a comment like this:

`# ()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()`

and repeat it several times.

**Comment by chrisbra** [comment]
I am not sure if the following is a good idea or not and I haven't been able to test it...
```diff
index 52ac48cbf..fac9f8685 100644
--- a/src/evalfunc.c
+++ b/src/evalfunc.c
@@ -10463,7 +10463,20 @@ do_searchpair(
            save_pos = curwin->w_cursor;
            curwin->w_cursor = pos;
            err = FALSE;
+#ifdef FEAT_RELTIME
+           proftime_T st;
+           profile_start(&st);
+#endif
            r = eval_expr_to_bool(skip, &err);
+#ifdef FEAT_RELTIME
+           profile_end(&st);
+           if (time_limit < st.tv_fsec
+# ifdef PROF_NSEC
+           * 1000
+# endif
+               )
+               err = TRUE;
+#endif
            curwin->w_cursor = save_pos;
            if (err)
            {
```

**Comment by yeti-dn** [comment]
I do not think it is a good idea – regardless whether it might be able to put plaster on the problem. It is the type of code just waiting to cause some weird behaviour later. 

The core issue is lack of error propagation. The functions higher in the chain do not know that something deeper in the chain has failed, whether due to a timeout or other reasons. ‘Negative’ return values are not enough. As we see here, if the failing thing happens to be a ‘skip’ function then negative result means including something that should be ignored.

But the second core issue is that it takes >150 ms to handle a kB of text in the first place. It seems to be a combination of interpreted vim code being slow with parsing the syntax over and over?
